### PR TITLE
CSP-258 initial code for reloading provider addresses

### PR DESCRIPTION
**Pull request by MarkFCain**

### CODE REVIEW

File: src/SFA.DAS.Roatp.Domain/Entities/ProviderAddress.cs [hunk 1]
```diff
@@ -16,5 +17,19 @@ public class ProviderAddress
...
+            {
+                ProviderId = source.ProviderId.GetValueOrDefault(),
+                AddressLine1 = source.Address1,
+                AddressLine2 = source.Address2,
+                AddressLine3 = source.Address3,
+                AddressLine4 = source.Address4,
+                Town=source.Town,
+                Postcode = source.Postcode,
+                AddressUpdateDate = DateTime.Now,
+                Id = source.Id
```
Comment (DevenKShah) [inline src/SFA.DAS.Roatp.Domain/Entities/ProviderAddress.cs:32]: Id is identity column , so why are we reading it from Ukrlp?

File: src/SFA.DAS.Roatp.Domain/Models/UkrlpProviderAddress.cs [hunk 2]
```diff
@@ -0,0 +1,15 @@
+﻿namespace SFA.DAS.Roatp.Domain.Models
+{
+    public class UkrlpProviderAddress
+    {
+        public int Id { get; set; }
```
Comment (DevenKShah) [inline src/SFA.DAS.Roatp.Domain/Models/UkrlpProviderAddress.cs:5]: There is no Id coming from outer api response is there?

File: src/SFA.DAS.Roatp.Jobs/Services/ILoadUkrlpAddressesService.cs [hunk 3]
```diff
@@ -0,0 +1,12 @@
+﻿using SFA.DAS.Roatp.Jobs.ApiModels.CourseDirectory;
+using System.Collections.Generic;
+using System.Threading.Tasks;
+using SFA.DAS.Roatp.Domain.Entities;
+
+namespace SFA.DAS.Roatp.Jobs.Services
+{
+    public interface ILoadUkrlpAddressesService
+    { 
+        public Task<bool> LoadUkrlpAddresses();
```
Comment (DevenKShah) [inline src/SFA.DAS.Roatp.Jobs/Services/ILoadUkrlpAddressesService.cs:10]: public access identifier unnecessary

File: src/SFA.DAS.Roatp.Jobs/Services/LoadUkrlpAddressesService.cs [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+            _importAuditWriteRepository = importAuditWriteRepository;
+            _logger = logger;
+        }
+
+        public async Task<bool> LoadUkrlpAddresses()
+        {
+            var timeStarted = DateTime.UtcNow;
+            var providers = await _providersReadRepository.GetAllProviders();
+      
+            var ukprnsSubset = providers.Select(provider => provider.Ukprn).Select(conv => (long)conv).ToList();
```
Comment (DevenKShah) [inline src/SFA.DAS.Roatp.Jobs/Services/LoadUkrlpAddressesService.cs:37]: Should filter down to providers with no address here.

File: src/SFA.DAS.Roatp.Jobs/Services/LoadUkrlpAddressesService.cs [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+                ukrlpProvider.ProviderId = providers.FirstOrDefault(x => x.Ukprn == ukrlpProvider.Ukprn)?.Id;
+                if (ukrlpProvider.ProviderId != null)
+                    providerAddresses.Add(ukrlpProvider);
+                else
+                {
+                    _logger.LogInformation($"There was no matching ProviderId for ukprn {ukrlpProvider.Ukprn}, so this was not added to ProviderAddress");
+                }
+            }
+
+            await _providerAddressesRepository.ReloadProviderAddresses(providerAddresses);
```
Comment (DevenKShah) [inline src/SFA.DAS.Roatp.Jobs/Services/LoadUkrlpAddressesService.cs:64]: Use bulk insert as we would have filtered down to providers with no addresses.

File: src/SFA.DAS.Roatp.Data/Repositories/ReloadProviderAddressesRepository.cs [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+        public async Task BulkInsert<T>(IList<T> data) where T : class
+        {
+            using var tx = await _roatpDataContext.Database.BeginTransactionAsync();
+
+            await _roatpDataContext.BulkInsertAsync(data);
+
+            await tx.CommitAsync();
+        }
+
+        public async Task<bool> ReloadProviderAddresses(List<ProviderAddress> providerAddresses)
```
Comment (DevenKShah) [inline src/SFA.DAS.Roatp.Data/Repositories/ReloadProviderAddressesRepository.cs:34]: I think deleting is not necessary, we can reload addresses where they are missing. So this method is redundant and you can use the BulkInsert method above which is not exposed via the interface yet.

File: src/SFA.DAS.Roatp.Data/Repositories/ReloadProviderAddressesRepository.cs [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+        private readonly RoatpDataContext _roatpDataContext;
+        private readonly ILogger<ReloadProviderAddressesRepository> _logger;
+
+        public ReloadProviderAddressesRepository(RoatpDataContext roatpDataContext, ILogger<ReloadProviderAddressesRepository> logger)
+        {
+            _roatpDataContext = roatpDataContext;
+            _logger = logger;
+        }
+
+        public async Task BulkInsert<T>(IList<T> data) where T : class
```
Comment (DevenKShah) [inline src/SFA.DAS.Roatp.Data/Repositories/ReloadProviderAddressesRepository.cs:25]: Unused?

File: src/SFA.DAS.Roatp.Domain/Entities/ProviderAddress.cs [hunk 8]
```diff
@@ -16,5 +17,19 @@ public class ProviderAddress
         public double? Longitude { get; set; }
         public DateTime? AddressUpdateDate { get; set; }
         public DateTime? CoordinatesUpdateDate { get; set; }
+
+        public static implicit operator ProviderAddress(UkrlpProviderAddress source)
+            => new ProviderAddress
+            {
+                ProviderId = source.ProviderId.GetValueOrDefault(),
```
Comment (DevenKShah) [inline src/SFA.DAS.Roatp.Domain/Entities/ProviderAddress.cs:24]: ProviderId is not coming from source is it? This we derive afterwards.

File: src/SFA.DAS.Roatp.Domain/Models/UkrlpProviderAddress.cs [hunk 9]
```diff
@@ -0,0 +1,15 @@
...
+        public int Ukprn { get; set; }
+        public string Address1 { get; set; }
+        public string Address2 { get; set; }
+        public string Address3 { get; set; }
+        public string Address4 { get; set; }
+        public string Town { get; set; }
+        public string Postcode { get; set; }
+        public int? ProviderId { get; set; }
```
Comment (DevenKShah) [inline src/SFA.DAS.Roatp.Domain/Models/UkrlpProviderAddress.cs:13]: We don't need providerId here, do we?

File: src/SFA.DAS.Roatp.Domain/Entities/ProviderAddress.cs (hunk 1 above)
Comment (MarkFCain) [reply src/SFA.DAS.Roatp.Domain/Entities/ProviderAddress.cs:32]: removed

File: src/SFA.DAS.Roatp.Jobs/Services/ILoadUkrlpAddressesService.cs (hunk 3 above)
Comment (MarkFCain) [reply src/SFA.DAS.Roatp.Jobs/Services/ILoadUkrlpAddressesService.cs:7]: removed

File: src/SFA.DAS.Roatp.Jobs/Services/LoadUkrlpAddressesService.cs (hunk 4 above)
Comment (MarkFCain) [reply src/SFA.DAS.Roatp.Jobs/Services/LoadUkrlpAddressesService.cs:37]: The story is to clear down all existing providerAddresses, and insert all new details.  So no filtering down here.

File: src/SFA.DAS.Roatp.Jobs/Services/LoadUkrlpAddressesService.cs (hunk 5 above)
Comment (MarkFCain) [reply src/SFA.DAS.Roatp.Jobs/Services/LoadUkrlpAddressesService.cs:64]: This is being done inside the providerAddressesRepository

File: src/SFA.DAS.Roatp.Data/Repositories/ReloadProviderAddressesRepository.cs (hunk 6 above)
Comment (MarkFCain) [reply src/SFA.DAS.Roatp.Data/Repositories/ReloadProviderAddressesRepository.cs:34]: The story is to delete all and re-insert, so I will leave this as is.  I've removed the unused BulkInsert method

File: src/SFA.DAS.Roatp.Domain/Entities/ProviderAddress.cs (hunk 8 above)
Comment (MarkFCain) [reply src/SFA.DAS.Roatp.Domain/Entities/ProviderAddress.cs:24]: I do inject the providerId into the 'source' earlier up the code, but it's a bit unintiuitive and smelly.  I think to make it less of a smell, I will insert it in a different way.  I did this to sidestep the issue that you cannot have more than one parameter in the implicit operator.

File: src/SFA.DAS.Roatp.Domain/Models/UkrlpProviderAddress.cs (hunk 9 above)
Comment (MarkFCain) [reply src/SFA.DAS.Roatp.Domain/Models/UkrlpProviderAddress.cs:13]: not any more

File: src/SFA.DAS.Roatp.Jobs/Services/LoadUkrlpAddressesService.cs [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+            _logger = logger;
+        }
+
+        public async Task<bool> LoadUkrlpAddresses()
+        {
+            var timeStarted = DateTime.UtcNow;
+            var providers = await _providersReadRepository.GetAllProviders();
+
+        
+            var ukprnsSubset = providers.Select(provider => provider.Ukprn).Select(conv => (long)conv).ToList();
```
Comment (DevenKShah) [inline src/SFA.DAS.Roatp.Jobs/Services/LoadUkrlpAddressesService.cs:38]: Does it have to be long? should we change the type in Entities instead?

File: src/SFA.DAS.Roatp.Jobs/Services/LoadUkrlpAddressesService.cs [hunk 11]
```diff
@@ -0,0 +1,89 @@
...
+
+            await _providerAddressesRepository.ReloadProviderAddresses(providerAddresses);
+
+            _logger.LogInformation("Provider addresses reload complete");
+            await _importAuditWriteRepository.Insert(new ImportAudit(timeStarted, providerAddresses.Count, ImportType.ProviderAddresses));
+
+            return true;
+        }
+
+        private static ProviderAddress MapProviderAddress(UkrlpProviderAddress source, int providerId)
```
Comment (DevenKShah) [inline src/SFA.DAS.Roatp.Jobs/Services/LoadUkrlpAddressesService.cs:73]: Why is this not an operator?

Comment (MarkFCain) [reply src/SFA.DAS.Roatp.Jobs/Services/LoadUkrlpAddressesService.cs:73]: needs 2 parameters.  An implicit operator takes 1 and only 1 parameter

Comment (MarkFCain) [reply src/SFA.DAS.Roatp.Jobs/Services/LoadUkrlpAddressesService.cs:73]: I need to map in the providerId for it to use BulkInsert

**Pull request merged**